### PR TITLE
Health-Based Acid Corrosion Overhaul

### PR DESCRIPTION
**Pull request by Kubius**
[REWORK]
## About the PR 

Massively retools the behavior of acid, primarily fluorosulfuric. Primary breakdown of changes:

- Fluorosulfuric acid damage on touch is now localized to five primary regions of the body - face, the rest of the head, the torso, the legs, and the arms. Protection, and damage, are calculated for all of these regions (the face takes extra damage compared to others), either based on fluid depth or across the upper body if splashed.
- Equipment now has an acid resilience property, which is automatically derived from the inverted permeability coefficient if it's lower than 1, up to a cap of 99%. This modifies the amount of damage the item's health takes from acid.
- Items that protect you from acid and aren't entirely immune to it now have a corrosion effect applied that gradually decays them over time. This corrosion effect factors in both the acid resilience property and any additional corrosion resistance from materials, and will visually melt objects with low acid resilience, like the melt status effect did.
- Items decaying from acid now have a durability bar, created only when that decay has occurred and appearing when the item is picked up or equipped. A precise number isn't given, only a color bar that indicates the general item health.
- Changeling fluorosulfuric acid has increased from 10 to 15 units to keep its damage stable against targets with only face exposed. As a consequence of the mechanics, damage is increased against targets with additional exposed skin.

## Why's this needed? 

Causes acid melting to happen on a more granular basis, with detectable progress and more recognition of which areas of the body are covered by equipment. This is pivotal for Nadir - #7730 - as the acid sea is intended to pose a threat that most gear can briefly shield you from, but no gear can permanently solve.

### CODE REVIEW

File: code/WorkInProgress/ObjectProperties.dm [hunk 1]
```diff
@@ -486,6 +486,16 @@ to say if there's demand for that.
...
+/datum/objectProperty/equipment/acidres
+	name = "Acid Resilience"
+	id = "acidres"
+	desc = "Slows corrosion. Influenced by permability."
+	tooltipImg = "acidres.png"
+	defaultValue = 10
+	getTooltipDesc(var/obj/propOwner, var/propVal)
+		return "[propVal]%"
+
+	ASSOCIATE_ATOM_PROPERTY(PROP_MOB_ACIDRES)
```
Comment (Tarmunora) [inline code/WorkInProgress/ObjectProperties.dm:498]: unused

File: _std/macros/atom_properties.dm [hunk 2]
```diff
@@ -213,6 +213,7 @@ To remove:
 #define PROP_MOB_EXPLOPROT(x) x("exploprot", APPLY_ATOM_PROPERTY_SUM, REMOVE_ATOM_PROPERTY_SUM)
 #define PROP_MOB_DISARM_RESIST(x) x("disarm_resist", APPLY_ATOM_PROPERTY_SUM, REMOVE_ATOM_PROPERTY_SUM)
 #define PROP_MOB_REFLECTPROT(x) x("reflection", APPLY_ATOM_PROPERTY_SIMPLE, REMOVE_ATOM_PROPERTY_SIMPLE)
+#define PROP_MOB_ACIDRES(x) x("acidres", APPLY_ATOM_PROPERTY_SUM, REMOVE_ATOM_PROPERTY_SUM)
```
Comment (Tarmunora) [inline _std/macros/atom_properties.dm:216]: unused

File: code/obj/item.dm [hunk 3]
```diff
@@ -882,6 +906,47 @@
...
+	if (src.corrosion > 0)
+		var/acid_taken_mult = 1
+		if(src.hasProperty("acidres")) //multiplier from permeability coefficient
+			acid_taken_mult = 0.01 * (100 - src.getProperty("acidres"))
+		if (istype(src.material) && src.material.hasProperty("corrosion")) //multiplier from corrosion resistance
+			acid_taken_mult = acid_taken_mult * (0.01 * (100 - src.material.getProperty("corrosion")))
+
+		var/base_melt = max(0.1 * src.corrosion,1) //calculate how much corrosion to consume, with a simple floor.
+		src.corrosion -= base_melt //consume the corrosion accordingly.
+		src.health -= 0.1 * base_melt * acid_taken_mult //apply the mult at this phase, so more acid resistance doesn't just prolong damage
```
Comment (Tarmunora) [inline code/obj/item.dm:920]: back-of-the-envelope math suggests that, when kept constantly at maximum corrosion, it takes a solid 2 minutes to melt through a space helmet. Which kinda makes changeling spit pretty worthless

File: code/modules/chemistry/Chemistry-Reagents.dm [hunk 4]
```diff
@@ -134,8 +134,11 @@ datum
 							for(var/atom in M.get_equipped_items())
 								if (istype(atom, /obj/item/clothing))
 									var/obj/item/clothing/C = atom
-									modifier -= (1 - C.permeability_coefficient)/3
-
+									if(C.item_function_flags & IMMUNE_TO_ACID)
+										modifier -= 0.5
```
[the remark this answers is not in the document]
Comment (Tarmunora) [reply code/modules/chemistry/Chemistry-Reagents.dm:138]: Also appears to not really be related to this PR at all

**Review by Tarmunora — CHANGES_REQUESTED** [review]
overall, I think this is a neat concept, but I don't think I'm a huge fan of it, at least not for general gameplay (i.e.: current station maps/modes/etc). Of note, the corrosion process having no recourse but to wait it out, and the degradation being irreversible save by getting entirely new equipment is not a gameplay loop that I think I'd want to see in the game as it stands.

File: code/modules/chemistry/Chemistry-Reagents.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kubius) [reply code/modules/chemistry/Chemistry-Reagents.dm:138]: Artifact of experimentation with reagent application mechanics. Will undo.

File: code/WorkInProgress/ObjectProperties.dm (hunk 1 above)
Comment (Kubius) [reply code/WorkInProgress/ObjectProperties.dm:498]: Tooltip desc is used, just not by name, if that's what you're referring to here - buildTooltipContent iterates over these.

File: _std/macros/atom_properties.dm (hunk 2 above)
Comment (Kubius) [reply _std/macros/atom_properties.dm:216]: Would argue this should exists for completeness' sake, but not entirely opposed to removing it.

**Comment by Kubius** [comment]
Potential changes:

- Allow corrosion process to be halted by washing items.
- Move this new corrosion functionality into a separate acid specifically for Nadir (to potentially later be backported into other acids once the feature balance has matured).

File: code/WorkInProgress/ObjectProperties.dm (hunk 1 above)
Comment (Tarmunora) [reply code/WorkInProgress/ObjectProperties.dm:498]: rather, the atom property is unused

**Comment by Kubius** [comment]
Changed in 3c953b8:

- Retooled touch calculation to now properly apply regional damage and decay based on fluid pool depth.
- Any source of cleaning that cleans off blood will now also stop items corroding (immediate stop if washing an individual item, or -15 corrosion if washing an entire mob).
- Dramatically increased the damage corrosion deals (1 point of base damage per 1 corrosion, which also correlates to 1 unit of acid right now). A notable breakpoint is that a 15-unit splash (i.e. ling acid spit) will now fully destroy a space helmet if it isn't promptly cleaned of acid.
- Increased the emergency hood's health to 5 to give it a brief window of surviving acid immersion.
- Fiddled with the commented-out section in fluid_core.dm. Section was (and still is) commented out because not doing the reagent reaction prevents gear decay from being calculated.

File: code/obj/item.dm (hunk 3 above)
Comment (Kubius) [reply code/obj/item.dm]: Rectified.

**Comment by Kubius** [comment]
The value caps adjusted in addcb48 to compensate for drastically increased corrosion damage: Maximum corrosion accumulated at once is now 30, from 40, and acid resilience cap is now 99%, from 98%.

File: code/obj/item.dm [hunk 5]
```diff
@@ -882,6 +913,47 @@
...
+	if (src.corrosion > 0)
+		var/acid_taken_mult = 1
+		if(src.hasProperty("acidres")) //multiplier from permeability coefficient
+			acid_taken_mult = 0.01 * (100 - src.getProperty("acidres"))
+		if (istype(src.material) && src.material.hasProperty("corrosion")) //multiplier from corrosion resistance
+			acid_taken_mult = acid_taken_mult * (0.01 * (100 - src.material.getProperty("corrosion")))
+
+		var/base_melt = max(0.1 * src.corrosion,1) //calculate how much corrosion to consume, with a simple floor.
+		src.corrosion -= base_melt //consume the corrosion accordingly.
+		src.health -= base_melt * acid_taken_mult //apply the mult at this phase, so more acid resistance doesn't just prolong damage
```
Comment (Tarmunora) [inline code/obj/item.dm:927]: And now if I splash a security officer with a single 10u click from a beaker (or a thrown glass) filled with pacid, and they have less than 10 seconds to get to a shower before they lose their secvest *and jumpsuit* (dropping their PDA/id, belt, and pocket items on the floor). If there is a balance to be struck here, this ain't it.

**Review by Tarmunora — CHANGES_REQUESTED** [review]

File: code/obj/item.dm (hunk 5 above)
Comment (Tarmunora) [reply code/obj/item.dm:927]: The key takeaway here imo is that pacid was balanced around only taking away mask/hat, whereas this change attempts to delete... basically all your clothes at once.

Comment (Kubius) [reply code/obj/item.dm:927]: Corresponding takeaway: Acid resilience from permeability probably won't work.

**Comment by Kubius** [comment]
Too many issues. Can't reconcile them.

**Comment by Kubius** [comment]
May reopen later.

**Pull request closed without merging**